File: projects/engineering.purdue.edu/kak/programmingwithobjects/pwocode/chap18/ThreadsBasicWithJoin.java
// This code example is from the following source:
//
// Book Title:  Programming with Objects, A Comparative Presentation
//              of Object-Oriented Programming with C++ and Java
//
// Book Author:   Avinash Kak
//
// Chapter:     Chapter 18  Multithreaded Object-Oriented Programming
//
// Section:     Section 18.1  Creating And Executing Simple Threads In Java
//



//ThreadsBasicWithJoin.java

class HelloThread extends Thread { 
    String message;

    HelloThread( String message ) { this.message = message; }

    public void run() {
        //int sleeptime = (int) ( Math.random() * 3000 );
        //try {                                 
        //    sleep( sleeptime );
        //  } catch( InterruptedException e ){}   
        System.out.print( message );
    }

    public static void main(String[] args) throws InterruptedException
    {
        HelloThread ht1 = new HelloThread( "Good" );
        HelloThread ht2 = new HelloThread( " morning" ); 
        HelloThread ht3 = new HelloThread( " to" );

        ht1.start();
        ht2.start();
        ht3.start();

        ht1.join();                                               //(H) 
        ht2.join();                                               //(I)
        ht3.join();                                               //(J)

        System.out.println( " you!" );
    }
}

File: projects/engineering.purdue.edu/scriptingwithobjects/swocode/chap14/ThreadsBasicWithJoin.py
#!/usr/bin/env python

###  ThreadsBasicWithJoin.py

import threading                                                     #(A)

class HelloThread( threading.Thread ):                               #(B)
    def __init__( self, message ):                                   #(C)
        self.message = message                                       #(D)
        threading.Thread.__init__(self)                              #(E)

    def run( self ):                                                 #(F)
#       for i in range(1000000): pass                                #(G)
        print self.message,                                          #(H)

t1 = HelloThread( "Good" )                                           #(I)
t2 = HelloThread( "Morning" )                                        #(J)
t3 = HelloThread( "to" )                                             #(K)

map( lambda x: x.start(), (t1, t2, t3) )                             #(L)

map( lambda x: x.join(), (t1, t2, t3) )                              #(M)

print "you!\n"                                                       #(N)
